consensus/src/block_storage/block_store.rs

```diff
@@ -96,6 +96,40 @@ pub struct BlockStore {
     time_service: Arc<dyn TimeService>,
 }
 
+pub fn update_counters_and_prune_blocks(
+    block_tree: Arc<RwLock<BlockTree>>,
+    storage: Arc<dyn PersistentLivenessStorage>,
+    root: Arc<ExecutedBlock>,
+    blocks_to_commit: &[Arc<ExecutedBlock>],
+    block_to_commit: Arc<ExecutedBlock>,
+) {
+    update_counters_for_committed_blocks(blocks_to_commit);
+    let current_round = root.round();
+    let committed_round = block_to_commit.round();
+    debug!(
+        LogSchema::new(LogEvent::CommitViaBlock).round(current_round),
+        committed_round = committed_round,
+        block_id = block_to_commit.id(),
+    );
+    event!("committed",
+        "block_id": block_to_commit.id().short_str(),
+        "epoch": block_to_commit.epoch(),
+        "round": committed_round,
+        "parent_id": block_to_commit.parent_id().short_str(),
+    );
+
+    let id_to_remove = block_tree.read().find_blocks_to_prune(block_to_commit.id());
+    if let Err(e) = storage.prune_tree(id_to_remove.clone().into_iter().collect()) {
+        // it's fine to fail here, as long as the commit succeeds, the next restart will clean
+        // up dangling blocks, and we need to prune the tree to keep the root consistent with
+        // executor.
+        error!(error = ?e, "fail to delete block");
+    }
+    block_tree
+        .write()
+        .process_pruned_blocks(block_to_commit.id(), id_to_remove);
```

Comment:
this shouldn't be delayed, this moves the consensus root, we should only delay the storage pruning.

Reply:
resolved